python/mxnet/io.py

```diff
@@ -411,10 +409,16 @@ def _getdata(self, data_source):
         """Load data from underlying arrays, internal use only"""
         assert(self.cursor < self.num_data), "DataIter needs reset."
         if self.cursor + self.batch_size <= self.num_data:
-            return [array(x[1][self.cursor:self.cursor+self.batch_size]) for x in data_source]
+            if isinstance(data_source[0][1], NDArray):
+                return [x[1][self.cursor:self.cursor+self.batch_size] for x in data_source]
+            else:
+                return [array(x[1][self.cursor:self.cursor+self.batch_size]) for x in data_source]
         else:
             pad = self.batch_size - self.num_data + self.cursor
-            return [array(np.concatenate((x[1][self.cursor:], x[1][:pad]),
+            if isinstance(data_source[0][1], NDArray):
+                return [concatenate([x[1][self.cursor:], x[1][:pad]]) for x in data_source]
+            else:
+                return [array(np.concatenate((x[1][self.cursor:], x[1][:pad]),
```

Comment:
same here. Change to
`return [concatenate([array(x[1][self.cursor:]), array(x[1][:pad])]) for x in data_source]`